6_visualize/src/prep_spark_line_fun.R

```diff
@@ -1,63 +1,184 @@
-
-      # Fill values lower than flood stage with the stage & then create a polygon out of it
-      above_flood_data <- mutate(storm_data_i, stage_normalized = pmax(stage_normalized, flood_stage_va))
-      above_flood_polygon <- data.frame(dateTime = head(above_flood_data$dateTime, 1),
-                                        stage_normalized = flood_stage_va) %>%
-        bind_rows(select(above_flood_data, dateTime, stage_normalized)) %>%
-        bind_rows(data.frame(dateTime = tail(above_flood_data$dateTime, 1),
-                             stage_normalized = min(above_flood_data$stage_normalized)))
-
-      y_coords <- c(y_pos, y_pos + vertical_spacing) # y coords relative to plot box
-      fig.new <- c(grconvertX(x_coords, from="npc", to="ndc"), grconvertY(y_coords, from="npc", to="ndc"))
-
-      op <- par(fig=fig.new, cex=2, new=TRUE, mar=rep(0, 4)) # setup parameters
-      # setup a plotting device
-      plot(x=NA, y=NA, type='n', axes=FALSE, xlab="", ylab="",
-           xlim = c(as.POSIXct(date_config$start, tz = "UTC"), as.POSIXct(date_config$end, tz = "UTC")),
-           ylim = c(0, 1)) # we assume "normalized" stage is between 0 and 1
-      # put stage polygons on plot
-      polygon(full_polygon$dateTime, full_polygon$stage_normalized, col = gage_col_config$gage_norm_col, border=NA)
-      polygon(above_flood_polygon$dateTime, above_flood_polygon$stage_normalized, col = gage_col_config$gage_flood_col, border=NA)
-      points(full_polygon$dateTime[-nrow(full_polygon)], full_polygon$stage_normalized[-nrow(full_polygon)],
-             col = gage_col_config$gage_flood_col, type='l', lwd=2.5)
-      par(op) # reset plot parameters
-
-      y_pos <- y_pos + vertical_spacing # increment spacing for next plot
+  # Compute normalized plot coordinates for all sites
+  x_coords <- c(spark_config$xleft, spark_config$xright)
+  vertical_spacing <- (spark_config$ytop - spark_config$ybottom) / length(sites)
+  y_coords <- data_frame(
+    site_no=sites,
+    lower=seq(spark_config$ytop - vertical_spacing, spark_config$ybottom, length.out=length(sites)),
+    upper=lower + vertical_spacing)
+
+  # Define functions that will convert each site's shapes to user coordinates
+  dateTime_to_x <- function(dateTime, x_user) {
+    date_lims_num <- as.numeric(date_lims, units='days')
+    date_time_num <- as.numeric(dateTime, units='days')
+    x_frac <- (date_time_num - date_lims_num[1]) / diff(date_lims_num)
+    x_user[1] + x_frac*diff(x_user)
+  }
+  stage_to_y <- function(stage_normalized, y_user) {
+    y_frac <- stage_normalized # stage_normalized is already a fraction between 0 and 1
+    y_user[1] + y_frac*diff(y_user)
+  }
+
+  # Prepare a list of shapes for each site: two polygons and one line
+  shapes <- list()
+  for(site in sites) {
+
+    # Filter data to just one site & create polygon out of it
+    storm_data_i <- filter(this_spark, site_no == site) %>%
+      arrange(dateTime) %>%
+      sf::st_set_geometry(NULL)
```

Comment:
I'm just making storm_data_i be a data_frame rather than an sf object. I don't remember the details, but I'm pretty sure that some of the manipulations below either weren't working or were creating objects larger than necessary when I kept storm_data_i as an sf object.